feat(backend): parameterize retryStrategy  input in Argo workflow

alyssacgoins opened the pull request:
**Description of your changes:**
This PR improves how the retryStrategy field is set in the argo workflow spec. When a component with setRetry enabled is included in a pipeline, two single separate container and executor templates are added - retry-system-container-executor and retry-system-container-impl. All four retryStrategy fields (duration, factor maxDuration and limit) are input as parameters into these templates, in order to allow the reuse (previously, separate templates were created for every unique retryStrategy configuration).

If the input pipeline contains no components with setRetry enabled, these two templates will not be created. 


**Checklist:**
- [x ] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -580,6 +637,26 @@ func extendPodMetadata(
 	}
 }
 
+func (c *workflowCompiler) getTaskRetryStrategyFromInput(maxCount string, backOffDuration string, backOffPolicy string,
+	backOffMaxDuration string) *wfapi.RetryStrategy {
+	backoff := &wfapi.Backoff{
+		Factor: &intstr.IntOrString{
+			Type:   intstr.String,
+			StrVal: backOffPolicy,
+		},
+		MaxDuration: backOffMaxDuration,
+		Duration:    backOffDuration,
+	}
```
```suggestion
func (c *workflowCompiler) getTaskRetryStrategyFromInput(maxCount string, backOffDuration string, backOffFactor string,
	backOffMaxDuration string) *wfapi.RetryStrategy {
	backoff := &wfapi.Backoff{
		Factor: &intstr.IntOrString{
			Type:   intstr.String,
			StrVal: backOffFactor,
		},
		MaxDuration: backOffMaxDuration,
		Duration:    backOffDuration,
	}
```

alyssacgoins replied:
Yep that variable naming makes much more sense

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -325,11 +349,18 @@ func (c *workflowCompiler) addContainerExecutorTemplate(name string, refName str
-		Name:          nameContainerImpl,
-		RetryStrategy: c.getTaskRetryStrategy(name),
+		Name: nameContainerImpl,
+		RetryStrategy: c.getTaskRetryStrategyFromInput(inputParameter(paramRetryMaxCount),
```
since the `getTaskRetryStrategy(name)` function is now unused, we should remove it

alyssacgoins replied:
Removed :+1:

alyssacgoins commented:
> Thanks for the PR @alyssacgoins ! Looks like there's some extra merge and dev commits in the history - Could you please rebase/squash your commits into a single (or more, if logical) commit(s)? Thanks!

Yep, squashed down to one commit :+1:

alyssacgoins commented on backend/src/v2/compiler/argocompiler/argo.go:
```diff
@@ -331,21 +331,25 @@ func hashValue(value interface{}) (string, error) {
+	paramPodSpecPatch            = "pod-spec-patch"             // a strategic patch merged with the pod spec
+	paramCondition               = "condition"                  // condition = false -> skip the task
+	paramKubernetesConfig        = "kubernetes-config"          // stores Kubernetes config
+	paramRetryMaxCount           = "retry-max-count"            // retry max count
```
Added some more in-detail comments - is that a good level of info?

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -257,11 +257,25 @@ func (c *workflowCompiler) containerExecutorTask(name string, inputs containerEx
+				Value: wfapi.AnyStringPtr(inputs.cachedDecision), Default: wfapi.AnyStringPtr("false"),
+			}, {
+				Name:  paramRetryMaxCount,
+				Value: wfapi.AnyStringPtr(c.getTaskRetryParameterValue(name, paramRetryMaxCount)),
```
Optional: I think it'd be cleaner if instead of `getTaskRetryParameterValue` there was a method that returns a slice of `wfapi.Parameter` with all of the retry parameters that are set. This means it could return an empty slice if `set_retry` wasn't called and we can then rely on default values in the arguments/parameters. This would also simplify the code but not needing to have a switch statement on the parameter name.

If you did this approach, I imagine the parameters would added like so:

```go
task.Arguments = append(task.Arguments, c.getTaskRetryParameterValues(name)...)
```

alyssacgoins replied:
Added - currently deciding how to refactor further.

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -292,6 +301,10 @@ func (c *workflowCompiler) addContainerExecutorTemplate(name string, refName str
 			Parameters: []wfapi.Parameter{
 				{Name: paramPodSpecPatch},
 				{Name: paramCachedDecision, Default: wfapi.AnyStringPtr("false")},
+				{Name: paramRetryMaxCount},
```
Could you set the defaults here so that the task doesn't need to set them and we can just always rely on the template having the correct defaults?

alyssacgoins replied:
Updated

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -517,38 +548,30 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
-			IntVal: retryPolicy.MaxRetryCount,
-		},
-		Backoff: backoff,
+	value := "0"
```
It seems like `paramRetryMaxCount` should default to `1`. Is there special behavior if it's set to `0`?

alyssacgoins replied:
Initially I was concerned with defaulting this value to 1, because I was setting a retryStrategy on every single template regardless of whether retry parameters were input - and a default "1" value would potentially cause every single template to retry. However, now that I'm using a 2-template approach, (retry and non-retry) there is no chance of this happening, so a default of "1" makes more sense - I have updated this.

mprahl commented:
@alyssacgoins have you tested the case where a pipeline fails and `set_retry` isn't set on the component? I'm wondering how the default parameter values of `"0"` are handled by Argo Workflows. Looking at the Argo Workflows code, it executes a different code path if retryStrategy is not nil and since default values are being passed, it is no longer nil due to Argo Workflows usage of pointers on values and thus `omitempty` on JSON doesn't take effect if a default Go value is passed (e.g. "" for string or 0 for int).

mprahl commented:
I think we should have at least one integration test that uses `set_retry` with all the options to ensure Argo Workflows accepts the values passed in as parameters by the KFP Argo compiler.

alyssacgoins commented:
> @alyssacgoins have you tested the case where a pipeline fails and `set_retry` isn't set on the component? I'm wondering how the default parameter values of `"0"` are handled by Argo Workflows. Looking at the Argo Workflows code, it executes a different code path if retryStrategy is not nil and since default values are being passed, it is no longer nil due to Argo Workflows usage of pointers on values and thus `omitempty` on JSON doesn't take effect if a default Go value is passed (e.g. "" for string or 0 for int).

I added this test case (samples/v2/pipeline_failure_no_retry.py) - though now I'm wondering if it's necessary because I implemented both a retry-enabled and non-retry-enabled template (therefore retryStrategy would be nil on the template produced by this test)

alyssacgoins commented:
> I think we should have at least one integration test that uses `set_retry` with all the options to ensure Argo Workflows accepts the values passed in as parameters by the KFP Argo compiler.

Added - samples/v2/pipeline_with_retry_all_args.py

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -257,10 +263,11 @@ func (c *workflowCompiler) containerExecutorTask(name string, inputs containerEx
-				{Name: paramCachedDecision, Value: wfapi.AnyStringPtr(inputs.cachedDecision), Default: wfapi.AnyStringPtr("false")},
-			},
+				{Name: paramCachedDecision, Value: wfapi.AnyStringPtr(inputs.cachedDecision), Default: wfapi.AnyStringPtr("false")}},
+				*c.getTaskRetryParameters(name, templateParameterInstructions{false, false, true})...,
```
This is the only time that `setValue` is set and `setDefault` and `setInputPath` aren't utilized in that code path. For readability, let's include the `setValue` branch directly in in the `containerExecutorTask` method or break it out to its own function.

alyssacgoins replied:
Created separate function `getTaskRetryParametersWithValue()` to handle setValue logic.  I believe this was a cleaner approach than including the setValue logic directly within the  `containerExecutorTask` method, but let me know what you think.

mprahl commented on samples/v2/sample_test.py:
```diff
@@ -37,6 +37,7 @@
 from kubernetes import client, config, utils
 import yaml
 from modelcar import modelcar
+from samples.v2 import pipeline_with_retry, pipeline_with_retry_all_args, pipeline_with_zero_retry
```
I think this will fix your CI failure.
```suggestion
import pipeline_with_retry, pipeline_with_retry_all_args, pipeline_with_zero_retry
```

alyssacgoins replied:
Thanks for pointing that out, updated :+1:

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,55 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+			parameters[i].Value = wfapi.AnyStringPtr(inputParameter(parameters[i].Name))
+		}
+		// If specified, set "Default" field with default parameter value.
+		if setDefault {
+			parameters[i].Default = wfapi.AnyStringPtr("0")
+			if parameterType == paramRetryMaxCount {
+				parameters[i].Default = wfapi.AnyStringPtr("1")
+			}
+		}
```
It seems like getTaskRetryParameters is also setting defaults. I'd prefer either of these options:

1. The parameters are always required and just handle setting default values in `getTaskRetryParameters`.
2. Modify `getTaskRetryParameters` to not return a parameter if it wasn't set in the task.

alyssacgoins replied:
I went with option 1 - if I understood correctly - I separated the logic for setting input parameter path and setting default value - they are each in separate functions now

mprahl commented on samples/v2/sample_test.py:
```diff
@@ -161,6 +162,9 @@ def test(self):
             TestCase(pipeline_func=parallel_after_dependency.loop_with_after_dependency_set),
+            TestCase(pipeline_func=pipeline_with_retry.retry_pipeline),
+            TestCase(pipeline_func=pipeline_with_retry_all_args.retry_pipeline),
+            TestCase(pipeline_func=pipeline_with_zero_retry.retry_pipeline),
```
How do these tests pass since `run_test_case` asserts that the test ran successfully?

alyssacgoins replied:
Facepalm moment here, I've updated the tests so that they don't raise an error (and now my build is passing too :slightly_smiling_face:)

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,55 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+		if setDefault {
+			parameters[i].Default = wfapi.AnyStringPtr("0")
+			if parameterType == paramRetryMaxCount {
+				parameters[i].Default = wfapi.AnyStringPtr("1")
```
I think you might have been right in your original choice that `0` is the correct default for this. I misread the code here in Argo Workflows:
```go
if retryStrategy.Limit != nil && limit != nil && int32(len(node.Children)) > *limit {
```

alyssacgoins replied:
Updated back to a single default value of 0

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
-	if backoffDuration != nil {
-		argoBackOffDuration = strconv.FormatInt(backoffDuration.Seconds, 10)
-	}
+func (c *workflowCompiler) getTaskRetryParametersWithValues(name string) *[]wfapi.Parameter {
```
We don't need to use a pointer here on the returned slice.
```suggestion
func (c *workflowCompiler) getTaskRetryParametersWithValues(name string) []wfapi.Parameter {
```

alyssacgoins replied:
Updated

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
 }
 
-func (c *workflowCompiler) getTaskRetryStrategy(name string) *wfapi.RetryStrategy {
+func (c *workflowCompiler) getTaskRetryParameters(name string) *[]wfapi.Parameter {
```
We don't need to use a pointer here on the returned slice.
```suggestion
func (c *workflowCompiler) getTaskRetryParameters(name string) []wfapi.Parameter {
```

alyssacgoins replied:
Updated

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
 	if retryPolicy == nil {
-		return nil
-	}
+		return &[]wfapi.Parameter{}
```
Optional: You can keep it as returning nil once you stop returning a slice pointer as a nil slice can still be read from (e.g. range over it) just not written to until it's initialized in Go.

alyssacgoins replied:
Makes sense, thanks for pointing that out - reverted back to returning nil

mprahl commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
-		},
-		MaxDuration: argoMaxDuration,
-		Duration:    argoBackOffDuration,
+func (c *workflowCompiler) addParameterDefault(parameters []wfapi.Parameter, defaultValue string) *[]wfapi.Parameter {
```
No need to return a pointer here.
```suggestion
func (c *workflowCompiler) addParameterDefault(parameters []wfapi.Parameter, defaultValue string) []wfapi.Parameter {
```

alyssacgoins replied:
Updated

mprahl commented on samples/v2/pipeline_with_retry.py:
```diff
@@ -0,0 +1,31 @@
+_KFP_PACKAGE_PATH = os.getenv('KFP_PACKAGE_PATH')
+
+@dsl.component(kfp_package_path=_KFP_PACKAGE_PATH)
+def print_op_retries(retries: str) -> str:
```
Optional: It doesn't really print anything anymore :laughing:

alyssacgoins replied:
Refactored this component to `verify_retries()` which returns a boolean value

mprahl reviewed (approved):
/lgtm

gmfrasca commented on samples/v2/pipeline_with_retry.py:
```diff
@@ -0,0 +1,30 @@
+if __name__ == '__main__':
+    compiler.Compiler.compile(
+        pipeline_func=retry_pipeline,
+        package_path='pipeline_with_retry.yaml')
```
code formatting/lint: add newline to end

gmfrasca commented on backend/src/v2/compiler/testdata/hello_world_with_retry_all_args.json:
```diff
@@ -0,0 +1,102 @@
+      }
+    }
+  }
+}
```
code formatting/lint: add newline to end

alyssacgoins replied:
Fixed. Applied this change to hello_world_with_retry.json as well

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+		return []wfapi.Parameter{}
+	}
+	// Create slice of all non-nil retryPolicy fields.
+	parameters := append([]wfapi.Parameter{},
+		wfapi.Parameter{Name: paramRetryMaxCount,
+			Value: wfapi.AnyStringPtr(retryPolicy.MaxRetryCount)})
```
```suggestion
	parameters := append(
	    []wfapi.Parameter{},
		wfapi.Parameter{
		    Name: paramRetryMaxCount,
			Value: wfapi.AnyStringPtr(retryPolicy.MaxRetryCount),
		},
	)
```

alyssacgoins replied:
Updated

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+			Value: wfapi.AnyStringPtr(retryPolicy.MaxRetryCount)})
+	if retryPolicy.GetBackoffDuration() != nil {
+		retryBackOffDuration := wfapi.AnyStringPtr(retryPolicy.GetBackoffDuration().Seconds)
+		parameters = append(parameters,
+			wfapi.Parameter{Name: paramRetryBackOffDuration,
+				Value: retryBackOffDuration})
```
```suggestion
		parameters = append(
		    parameters,
			wfapi.Parameter{
			    Name: paramRetryBackOffDuration,
				Value: retryBackOffDuration,
			},
		)
```

alyssacgoins replied:
Updated

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+			wfapi.Parameter{Name: paramRetryBackOffDuration,
+				Value: retryBackOffDuration})
+	}
+	parameters = append(parameters,
+		wfapi.Parameter{Name: paramRetryBackOffFactor,
+			Value: wfapi.AnyStringPtr(retryPolicy.GetBackoffFactor())})
```
```suggestion
	parameters = append(
	    parameters,
		wfapi.Parameter{
		    Name: paramRetryBackOffFactor,
			Value: wfapi.AnyStringPtr(retryPolicy.GetBackoffFactor()),
		},
	)
```

alyssacgoins replied:
Updated

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -527,38 +536,61 @@ func (c *workflowCompiler) getTaskRetryPolicySpec(name string) *pipelinespec.Pip
+			Value: wfapi.AnyStringPtr(retryPolicy.GetBackoffFactor())})
+	if retryPolicy.GetBackoffMaxDuration() != nil {
+		retryBackOffMaxDuration := wfapi.AnyStringPtr(retryPolicy.GetBackoffMaxDuration().Seconds)
+		parameters = append(parameters,
+			wfapi.Parameter{Name: paramRetryBackOffMaxDuration,
+				Value: retryBackOffMaxDuration})
```
```suggestion
		parameters = append(
		    parameters,
			wfapi.Parameter{
			    Name: paramRetryBackOffMaxDuration,
				Value: retryBackOffMaxDuration,
			},
		)
```

alyssacgoins replied:
Updated

gmfrasca commented on backend/src/v2/compiler/argocompiler/container.go:
```diff
@@ -289,8 +290,8 @@ func (c *workflowCompiler) addContainerExecutorTemplate(name string, refName str
 	if taskRetrySpec != nil {
-		nameContainerExecutor = name + "-" + nameContainerExecutor
-		nameContainerImpl = name + "-" + "system-container-impl"
+		nameContainerExecutor = "retry" + "-" + nameContainerExecutor
+		nameContainerImpl = "retry" + "-" + "system-container-impl"
```
```suggestion
		nameContainerExecutor = "retry-" + nameContainerExecutor
		nameContainerImpl = "retry-" + nameContainerImpl
```

alyssacgoins replied:
Updated

gmfrasca reviewed (commented):
/lgtm
